Allow loading XRC from wxXmlDocument

vadz opened the pull request:
I hesitated for a long time between `LoadDocument()` and `LoadFromStream()` and finally went with the former, but I'm still not quite sure if it's really the best. Does anybody see any arguments in favour of one or the other?

@vslavik Do you see anything wrong with this? I was surprised that `wxXmlResource` didn't support this already when it seems like a natural thing to do.

vadz committed:
Add wxXmlResource::LoadDocument()

This allows loading XRC from anywhere, not just files and URLs.

vadz committed:
Use wxXmlResource::LoadDocument() in XRC unit tests

This is simpler and better than creating a temporary file unnecessarily
and also serves as a test of the just added function.

vslavik commented:
> I hesitated for a long time between LoadDocument() and LoadFromStream() and finally went with the former, but I'm still not quite sure if it's really the best. 

I understand why "document", but don't understand "stream". IOW I think this is a (much) better name.

Wouldn't a better signature be `bool LoadDocument(wxXmlDocument* doc, const wxString& name = "");` — for uses where you don't care about unloading?

vslavik commented:
> OK, I'll keep this one, but the idea with "stream" was to just pass the wx{Memory,String}InputStream to the function

Ah, I misunderstood you as choosing between the *names*, which confused me. I understand the appeal of `LoadFromStream()`.

The argument in favour of doing it as this PR does would be that `LoadFromStream()` can be trivially implemented once you have `LoadDocument()` (as you show in the example), while the other way around would be jumping through inefficient hoops. 

And I agree this is useful to have, thanks for doing it!

vadz committed:
Don't try relative paths in wxFileSystem if current path is empty

This seems redundant, as the loop is the same as the one just below it
which is executed in any case if this one doesn't find anything.

vadz committed:
Micro optimize string concatenation in wxFileSystem::OpenFile()

Concatenate the string only once instead of doing it several times.
Compiler might be optimizing this anyhow in release builds, but it
definitely helps in at least the debug ones and doesn't cost anything.

vadz committed:
Document and test that bitmap paths in XRC are percent-encoded

These paths are actually URLs and so the special URL characters must be
percent-encoded in them.

Document this and add a test checking that this is how it works.

Closes #19142.

vadz commented:
Updated now and also added a couple of (very) minor optimizations and a test for the problem of [this ticket](https://trac.wxwidgets.org/ticket/19142) which is why I had initially started looking at this in the first place.
